keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+#### Story 1
+
+I'm an application developer building a Kubernetes controller that I want to deploy into my cluster.  I want to make sure that my controller securely authenticates to kube-apiserver using mTLS.
+
+I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `noAutomountServiceAccountToken` because I want to ensure that my pod cannot use bearer tokens for authentication.
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: pod-certificates-example
+spec:
+  restartPolicy: OnFailure
+  automountServiceAccountToken: false
+  containers:
+  - name: main
+    image: debian
+    command: ['sleep', 'infinity']
+    volumeMounts:
+    - name: kube-apiserver-client-certificate
+      mountPath: /var/run/secrets/kubernetes.io/api-mtls-access
+  volumes:
+  - name: kube-apiserver-client-certificate
+    projected:
+      sources:
+      - podCertificate:
+          signerName: "kubernetes.io/kube-apiserver-client-pod"
+          credentialBundlePath: client-credentials.pem
+      - configMap:
+          localObjectReference: kube-root-ca.crt
+          items:
+          - key: ca.crt
+            path: kube-apiserver-root-certificate.pem
+      - downwardAPI:
+          items:
+          - path: namespace
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+```
+
+I ensure that my pod uses client-go's InClusterConfig, and client-go automatically uses the mTLS client certificate to authenticate to kube-apiserver.
+
+### Risks and Mitigations
+
+**Scalability**: In the current design and draft implementation, Kubelet holds
+the private keys for each pod on the node in memory.  If Kubelet restarts, then
+the private keys are lost, and Kubelet will need to immediately generate new
+keys and request new certificates for each volume on the pod.  If Kubelet
+crashloops, this can generate high CertificateSigningRequest traffic on
+kube-apiserver.  This can potentially be mitigated by giving Kubelet the
+capability to read back the existing key from the projected volume on startup.
+
+## Design Details
+
+### X.509 PodIdentity Extension
+
+The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
+the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
+identity information into an X.509 certificate or certificate request, we can
+define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
+define a management policy for the PEN) and specify that it contains the
+following DER-encoded structure:
+
+```go
+type podIdentityASN1 struct {
```

Comment:
Somewhat off-topic, but... What would it take to network bind these certificates? E.g. assuming we can trust the underlying network to prevent IP spoofing, could we include a node or pod source IP here as well?

Reply:
Can it be checked as an *issuance* gate for the CSR request, as a pod attestation?

Or do we have to take the declared binding at face value, and check it at time-of-use only?

Reply:
Doesn't even have to be `kube-apiserver` that validates during admission, IIUC?

E.g. you could use SPIRE's attesting agent to validate the admission request for the cert and reject if out-of-band validations fail, or any other thing.